derive/src/lib.rs

```diff
@@ -101,12 +101,17 @@ fn generate_type(input: TokenStream2) -> Result<TokenStream2> {
         }
     });
 
-    let ast: DeriveInput = syn::parse2(input.clone())?;
+    let mut ast: DeriveInput = syn::parse2(input.clone())?;
     let build_type = match &ast.data {
         Data::Struct(ref s) => generate_composite_type(s, &scale_info),
         Data::Enum(ref e) => generate_variant_type(e, &scale_info),
         Data::Union(_) => return Err(Error::new_spanned(input, "Unions not supported")),
     };
+
+    // Remove any type parameter defaults, we don't want those. E.g. `impl<T: Stuff = WutEven>`
+    ast.generics.type_params_mut().for_each(|type_param| {
+        type_param.default = None;
+    });
```

Comment:
Is there a way to use `ImplGenerics` from `split_for_impl()` above to avoid  this? Not sure whether it will work with our modified lifetimes or not.

Reply:
You are correct that `ImplGenerics` takes care of [removing the defaults](https://docs.rs/syn/1.0.61/src/syn/generics.rs.html#1068) and it takes care of const parameter defaults too, so it's strictly better.

However there's another problem: we replace all lifetimes with `'static` and this means that given this input:

```
struct Goat<'bar, T: SomeTrait = SomeDefault>
```

 we get this in the expansion:

```
impl<'static, T: SomeTrait>
```

So `ImplGenerics` is the right tool for the job, but using it will require cloning&munging elsewhere. Not sure what is least worse tbh. :/

Reply:
I have an idea: remove the replacing of lifetimes and add something like this to `make_where_clause`:

```
for lifetime in generics.lifetimes() {
        where_clause
            .predicates
            .push(parse_quote!(#lifetime : 'static));
    }
```
That way we should be able to use `ImplGenerics` and remove completely the two mutable bits.

Reply:
You are completely right. And as an extra bonus we now don't need to clone the input anymore. :heart:

Reply:
With your suggestion we get an expansion like this:

```rust
    struct Assoc<'bar, T: Types> {
        a: T::A,
        b: &'bar u64,
    }
    
    const _: () = {
        impl<'bar, T: Types> ::scale_info::TypeInfo for Assoc<'bar, T>
        where
            'bar: 'static,
            T::A: ::scale_info::TypeInfo + 'static,
            T: Types + ::scale_info::TypeInfo + 'static,
			…
			…
```

That `'bar: 'static` looks a bit funny but it's valid syntax so all good.